Update IceCube Gold/Bronze alert example and schema

blaufuss opened the pull request:
more realistic example notice example, based on a recent alert on Classic.  

Maintain more uniform naming of alert stream from classic days.  
Update units for consistency across notice
add asymmetric errors for that we send in rev1+ of alerts (currently only on GCN circulars)

blaufuss commented on gcn/notices/icecube/GoldAndBronze.example.json:
```diff
@@ -1,54 +1,30 @@
-    }
+    "far": 8.029e-8
+  },
+  "asymmetric_errors": {
```
no, circular error from initial online alert, asymmetric rectangular box from offline reconstruction and in updated notice.  

Plan is to include a link to healpix URL on our website:  
"healpix_url": "https://roc.icecube.wisc.edu/public/"

blaufuss replied:
(healpix url will only be there with updated notices)

blaufuss replied:
Question:  for a rectangular error box, what should the "position_angle" be for a box aligned with RA/DEC?  Is RA or DEC the major axis in that case?  Apologies if I've missed the definition of this

Vidushi-GitHub replied:
Don't worry if the shape is not tilted or aligned with the standard (RA, Dec) axis, as the default value is set to 0. All axis measurements are taken from the center of the shape, hence, for a rectangle, the semi-major axis is half of the length. Please use the enum:"rectangle" and include asymmetric errors of length as follows: 'semi-major': [RA-, RA+] and 'semi-minor': [Dec-, Dec+].

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.schema.json:
```diff
@@ -2,11 +2,11 @@
+  "title": "IceCubeGoldAndBronzeTracks",
+  "description": "IceCube Astrotrack Gold And Bronze track alert events",
   "properties": {
     "type": {
```
We've recently updated the schema to a more human-readable format that uses fewer curly braces {}. Please update your schema by replacing 'properties' with 'allOf' (ex:https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/fermi/lat/Alert.schema.json).

blaufuss replied:
Thanks for the suggestion.  I'm far from a JSON schema expert, but doesn't allOf require that all items are provided all the time?  I think that will work here, as we don't have any items that will only appear conditionally.    

We're working on other schema, namely for LVK followup where some items will be conditionally included (say if there'a a neutrino in spatial and temporal coincidence with the LVK localization.  In these cases, allOf won't work as a replacement to properties and required?  Is there a preferred way to handle these?  thanks

blaufuss replied:
Additionally:  for the IceCube specific items in the schema, they should be locally defined in the schema file to include with allOf, defined in a separate file or look to make additions to the core area?

Vidushi-GitHub replied:
Yes, it is true that allOf requires all the schemas referred to. However, if you need more flexibility, for instance, to accommodate LVK localization, you can use anyOf (https://json-schema.org/understanding-json-schema/reference/combining.html#anyof) instead. We would suggest to use combination of allOf (for must schema sets) and anyOf (for optional schemas). I am also trying the JSON for the first time, thanks to @dakota002!

Regarding your additional question, we also had a discussion about it, we decided to define the Specific schema fields within the same schema instead of referring to a separate file. Thus, please go with defining the IceCube specific schema within alert schema, as already done, flexible structure would be: "allOf": [{$ref1}, {$ref2}...], "anyOf": [{$ref1}, {$ref2}...], "properties":{specific schema}.

blaufuss commented:
all requested changes should be addressed

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example_update.json:
```diff
@@ -0,0 +1,21 @@
+  "dec": 9.01,
+  "uncertainty_shape": "rectangle",
+  "position_angle": 0,
+  "semi_major": [-1.5, 1.4],
```
```suggestion
  "ra_axis": [-1.5, 1.4],
```

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example_update.json:
```diff
@@ -0,0 +1,21 @@
+  "uncertainty_shape": "rectangle",
+  "position_angle": 0,
+  "ra_axis": [-1.5, 1.4],
+  "semi_minor": [-1.2, 1.3],
```
```suggestion
  "dec_axis": [-1.2, 1.3],
```

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example.json:
```diff
@@ -1,51 +1,17 @@
+  "ra": 345.82,
+  "dec": 9.01,
+  "uncertainty_shape": "circle",
+  "semi_major": [0.51],
```
```suggestion
  "ra_axis": [0.51],
```

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example_update.json:
```diff
@@ -0,0 +1,21 @@
+  "containment_probability": 0.9,
+  "systematic_included": false,
+  "healpix_url": "https://roc.icecube.wisc.edu/public/",
+  "trigger_time": "2023-04-16T05:22:26.150574",
```
```suggestion
  "trigger_time": "2023-04-16T05:22:26.150574Z",
```

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example.json:
```diff
@@ -1,51 +1,17 @@
-  }
+  "additional_info": "IceCube Bronze Neutrino Track Alert",
+  "id": ["137840", "57034692"],
+  "datetime_of_notice": "2023-04-16T05:22:29.55",
```
```suggestion
  "datetime_of_notice": "2023-04-16T05:22:29.55Z",
```

dakota002 commented:
@blaufuss  For your issue with validation when using allOf, you can use a combination of allOf, anyOf, or oneOf to specify more exactly how the fields should be validated. 

These keywords behave as AND, OR, and XOR respectively:
- All items in an allOf must be valid
- One or more item in the anyOf must be valid
- Exactly one item in the oneOf must be valid

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example.json:
```diff
@@ -1,51 +1,17 @@
+  "ra_axis": [0.51],
+  "containment_probability": 0.9,
+  "systematic_included": false,
+  "trigger_time": "2023-04-16T05:22:26.150574",
```
```suggestion
  "trigger_time": "2023-04-16T05:22:26.150574Z",
```

Vidushi-GitHub commented on gcn/notices/icecube/GoldAndBronze.example_update.json:
```diff
@@ -0,0 +1,21 @@
+  "additional_info": "IceCube Bronze Neutrino Track Alert",
+  "event_name": ["IceCube-230416A"],
+  "id": ["137840", "57034692"],
+  "datetime_of_notice": "2023-04-16T05:42:00.0",
```
```suggestion
  "datetime_of_notice": "2023-04-16T05:42:00.0Z",
```

Vidushi-GitHub commented:
@blaufuss we have updated your schema with minor changes, like re-worded the semi_major to ra_axis, likewise semi_minor. Trigger time was reported as "2023-04-16T05:22:26.150574", added Z at the end.

blaufuss commented:
> @blaufuss we have updated your schema with minor changes, like re-worded the semi_major to ra_axis, likewise semi_minor. Trigger time was reported as "2023-04-16T05:22:26.150574", added Z at the end.

Thanks, anything else needed?

dakota002 commented:
> > @blaufuss we have updated your schema with minor changes, like re-worded the semi_major to ra_axis, likewise semi_minor. Trigger time was reported as "2023-04-16T05:22:26.150574", added Z at the end.
> 
> Thanks, anything else needed?

If everything looks good and you're validation is working as expected then I think we can go ahead and get this merged in. Also if my comment ^^^ about allOf/anyOf/oneOf makes sense and you need to do any restructuring we can wait for that too

blaufuss commented:
> > > @blaufuss we have updated your schema with minor changes, like re-worded the semi_major to ra_axis, likewise semi_minor. Trigger time was reported as "2023-04-16T05:22:26.150574", added Z at the end.
> > 
> > 
> > Thanks, anything else needed?
> 
> If everything looks good and you're validation is working as expected then I think we can go ahead and get this merged in. Also if my comment ^^^ about allOf/anyOf/oneOf makes sense and you need to do any restructuring we can wait for that too

Thanks.  Let's stick with anyOf for now if it's OK.  Can tweak as things settle and we're sure we're supplying all items in a core class, then we can break it out into a allOf list.  I suspect the example initial and update will be more useful for consumers...
